include/seqan3/io/alignment/sam_tag_dictionary.hpp

```diff
@@ -0,0 +1,401 @@
+template <> struct sam_tag_type<"R2"_tag> { using type = std::string; /*!< The type.*/ };
+//!\brief Specialization of seqan3::sam_tag_type defining type member **std::string**.
+template <> struct sam_tag_type<"RG"_tag> { using type = std::string; /*!< The type.*/ };
+//!\brief Specialization of seqan3::sam_tag_type defining type member **std::string**.
+template <> struct sam_tag_type<"RT"_tag> { using type = std::string; /*!< The type.*/ };
+//!\brief Specialization of seqan3::sam_tag_type defining type member **std::string**.
+template <> struct sam_tag_type<"RX"_tag> { using type = std::string; /*!< The type.*/ };
+
+// template <> struct sam_tag_type<"S2"_tag> {};
+
+//!\brief Specialization of seqan3::sam_tag_type defining type member **std::string**.
+template <> struct sam_tag_type<"SA"_tag> { using type = std::string; /*!< The type.*/ };
+//!\brief Specialization of seqan3::sam_tag_type defining type member **int32_t**.
+template <> struct sam_tag_type<"SM"_tag> { using type = int32_t; /*!< The type.*/ };
+
+// template <> struct sam_tag_type<"SQ"_tag> {};
+
+//!\brief Specialization of seqan3::sam_tag_type defining type member **int32_t**.
+template <> struct sam_tag_type<"TC"_tag> { using type = int32_t; /*!< The type.*/ };
+//!\brief Specialization of seqan3::sam_tag_type defining type member **std::string**.
+template <> struct sam_tag_type<"U2"_tag> { using type = std::string; /*!< The type.*/ };
+//!\brief Specialization of seqan3::sam_tag_type defining type member **int32_t**.
+template <> struct sam_tag_type<"UQ"_tag> { using type = int32_t; /*!< The type.*/ };
+//!\}
+
+/*!\brief The SAM tag dictionary class that stores all optional SAM fields.
+ * \ingroup sam_tag_dictionary
+ *
+ * \details
+ *
+ * The dictionary can be accessed via the functions seqan3::sam_tag_dictionary::get() and
+ * seqan3::sam_tag_dictionary::set(). Every time the SAM tag you wish to query
+ * for must be given as a template argument to the functions.
+ *
+ * Example:
+ *
+ * \snippet test/snippet/io/alignment/sam_tag_dictionary/general_usage.cpp all
+ *
+ * \attention You can get any SAM_tag out of the dictionary, even if the tag is
+ *            user defined, but note that for unkown tags the return type is an
+ *            [std::variant](https://en.cppreference.com/w/cpp/utility/variant).
+ *            If you want specify the return type of your custom tag, you need
+ *            to overload the seqan3::sam_tag_type metafunction.
+ *
+ * \snippet test/snippet/io/alignment/sam_tag_dictionary/unknown_tag.cpp all
+ *
+ * As mentioned before you can either overload the metafunction seqan3::sam_tag_type
+ * for the tag "XZ" or learn more about an std::variant at
+ * https://en.cppreference.com/w/cpp/utility/variant.
+ *
+ * \sa seqan3::sam_tag_type
+ * \sa https://en.cppreference.com/w/cpp/utility/variant
+ * \sa https://samtools.github.io/hts-specs/SAMv1.pdf
+ * \sa https://samtools.github.io/hts-specs/SAMtags.pdf
+ */
```

Comment:
In general I would prefer not to have `get` and `set`, but instead have `get` return by reference so that you can call:
```cpp
std::cout << dict.get<"NM"_tag>() << '\n';
dict.get<"NM"_tag>() = 7;
```

This would be in line with the other get functions from the standard library.

Reply:
And then overload the assignment oprator? Since I have some requirements on the type.

Reply:
why would you need to overload the assignment operator?

Reply:
Ok true the type requirements come with the get functionality,
but I have another problem: How can get<> return an lvalue reference if I extract the value from the variant inside the get function?

Reply:
std::get on a variant also returns an lvalue reference, that's the point of the variant, there is no conversion!